Kajabi/sage-lib
# SPIKE Combining old and new

## Description

**NOTE: This contains work from `develop` and `SAGE-19_EPIC-foundations` that is out of date and will need to be carefully synced with the latest from each branch before it can be merged.**

This PR is a draft of how we could combine the legacy and next system themes into a single system. Here are the changes summarized:

- We introduce the idea of "theme" for distinguishing between the current "legacy" look-and-feel versus the new, "next" look-and-feel.
- SageRails:
  - We introduce a global variable, `$sage_theme` that can be set in an application to determine which "theme" is returned by any calls to Sage components.
  - In calls to `sage_component` this variable is used to retrieve the desired theme's view. These views are completely separated so that the markup for each view can be managed cleanly independently. However, the main `.rb` files for each component are still unified so as to ensure a common API is retained. This will help ensure cleaner upgrade/downgrade for existing uses of the components.
  - The underlying token Dictionary is also switched up based on the `$sage_theme`. This impacts values available to `SageTokens` and `SageClassnames`.
  - In the documentation app, a toggle is implement in the UI to allow users/developers to actively toggle which theme is visible in the interface.
- `sage-assets`:
  - The entire stylesheet system is split into "themes" and compiling into separate `dist` files: `dist/main-legacy.scss` and `dist/main-next.scss` that can be loaded separately by consuming applications.
  - Direct access to `scss` files within `lib` will need to update to point to one or the other `theme` subfolder.
  - `lib/_consumables.scss` still exists but points to the `legacy` folder for now.
- `sage-react`:
  - The entire component library is split into "themes". However, all components are still exporting but dependenant on `window.SAGE_THEME`; by default, the "legacy" components are returned, but this global variable can be set to allow the "next" components to return in stead.
  - For local development, the Storybook start script launches two different storybook dev servers at 4100 (legacy) and 4110 (next) for separate development and testing. The `legacy` system is still configured as the public release for now.
- `sage-system`: no changes here. All work will need to ensure both component Rails component structures are supported.
- `sage-packs`: additional pack are configured to pull in the different themes.


## Testing in `sage-lib`

- See the Theme switcher in the top right of the docs site
- See localhost:4100 for legacy Storybook and localhost:4110 for next Storybook


## Testing in `kajabi-products`
(BREAKING): Introduces the new "next" theme alongside the current "legacy" theme. [This PR](https://github.com/Kajabi/kajabi-products/pull/23190) introduces the necessary changes in `kp` to implement this structure safely.

## Related

https://kajabi.atlassian.net/browse/SAGE-434
